### PR TITLE
Skater as an Overlay (no more popup)

### PR DESCRIPTION
**Pull request by Ecalzo**
This is a mostly-functioning version of Skater as a webpage overlay. This is being developed for the following reasons:
* The extension is too slow as a standard chrome/FF popup. This is much much faster.
* Being spotlight-inspired, this allows us to mimic the look/feel of Spotlight more closely

Core areas where I am looking for feedback:
* How should I organize this? I have not figured out how to break this into multiple console scripts, but it may be simpler than I assume.
* How should I test this thing? (kind of a TODO to learn about jest)
* Being much faster, what else would you like this extension to be able to search through? Any thoughts on where else this can fit into a workflow?

### CODE REVIEW

File: background.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    } else {
+      sendResponse([]);
+    }
+    // the trick was to return true from the addListener function, mind blown
+    return true
+  }
+);
+
+
+function searchBookmarks(query) {
```
Comment (wojm) [inline background.js]: no longer used?

File: overlay.js [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                    moveUpOneResult();
+                    return
+                }
+            case "Down":
+            case "ArrowDown":
+                    // move to next search result
+                    moveDownOneResult();
+                    return
+        }
+    } else { return }
```
Comment (wojm) [inline overlay.js]: seems a bit odd to do this, everything below can be included in above if block

File: overlay.js [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    } else if (getSearchInputElement()) {
+        const focusedElement = document.activeElement;
+        switch(documentEvent.key) {
+            case "Up":
+            case "ArrowUp":
+                // move to last search result or input
+                // FIXME functionize
+                if (focusedElement.getAttribute('class') === 'skater-link skater-result-0') {
+                    focusInput();
+                    return
```
Comment (wojm) [inline overlay.js]: `break` not `return`

File: overlay.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            resultsDiv.appendChild(createListItem(result, index));
+            index += 1;
+        });
+    } else {
+        ensureResultsListIsHidden();
+    };
+}
+
+
+function focusInput() {
```
Comment (wojm) [inline overlay.js:319]: focusSearchInputElement

File: overlay.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    const resultsDiv = getSearchResultsElement();
+    // wipes the unordered list
+    resultsDiv.innerHTML = '';
+    if (results.length) {
+        ensureResultsListIsVisible();
+        let index = 0;
+        results.forEach(result => {
+            resultsDiv.appendChild(createListItem(result, index));
+            index += 1;
+        });
```
Comment (wojm) [inline overlay.js:313]: results.forEach((result, index) => createListItem(result, index));

File: overlay.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+//         links.forEach(link => {
+//         link.addEventListener('click', event => {
+//             const ctrlPressed = (event.ctrlKey || event.metaKey);
+//             const url = event.target.href;
+//             // TODO: Decide if this is necessary
+//             chrome.tabs.create({'url': url, active: !ctrlPressed});
+//         }, false); 
+//     });
+// }
+// }
```
Comment (wojm) [inline overlay.js]: commented out code is useless. you can refer back to it in your commit history

File: overlay.js [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    searchResultsElement.style.padding = "0px";
+}
+
+function refineResults(searchResults, query) {
+    return searchResults.filter(result => {
+      // cuts bookmark Title down to a substring for closer matching
+      const queryLen = query.length;
+      const queryLower = query.toLowerCase();
+      const bookmarkTitle = result.title.substring(0, queryLen).toLowerCase();
+      return bookmarkTitle.includes(queryLower) & result.hasOwnProperty('url');
```
Comment (wojm) [inline overlay.js:299]: result.url

File: overlay.js [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    searchResultsElement.style.visibility = "hidden";
+    searchResultsElement.style.padding = "0px";
+}
+
+function refineResults(searchResults, query) {
+    return searchResults.filter(result => {
+      // cuts bookmark Title down to a substring for closer matching
+      const queryLen = query.length;
+      const queryLower = query.toLowerCase();
+      const bookmarkTitle = result.title.substring(0, queryLen).toLowerCase();
```
Comment (wojm) [inline overlay.js:298]: result.title.toLowerCase().startsWith(queryLen.toLowerCase())

File: overlay.js [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+function ensureResultsListIsVisible() {
+    const searchResultsElement = getSearchResultsElement();
+    searchResultsElement.style.visibility = "visible";
+    searchResultsElement.style.padding = "0px 10px 10px 10px";
+}
+
+function ensureResultsListIsHidden() {
+    const searchResultsElement = getSearchResultsElement();
+    searchResultsElement.style.visibility = "hidden";
+    searchResultsElement.style.padding = "0px";
```
Comment (wojm) [inline overlay.js:290]: weird you change padding

File: overlay.js [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+
+function createListItem(result, index) {
+    const listElement = document.createElement('div');
+    const listURL = document.createElement('a');
+    listURL.setAttribute('class', `skater-link skater-result-${index}`);
+    listURL.style.color = "#81b3d2";
+    listElement.setAttribute('class', `searchResultItem`);
+    listElement.style['padding-top'] = "5px";
+
+    if (result.title.length > 27) {
```
Comment (wojm) [inline overlay.js:269]: const MAX_RESULTS = 27; 

at the top

File: overlay.js [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+function createSearchResultsList() {
+    const resultsDiv = document.createElement('div');
+    resultsDiv.id = "searchResults";
+    resultsDiv.style = "padding: 20px; padding 20px; background-color: #f5f5f5; font-size:17px; font-family: Helvetica Neue,Helvetica,Arial,sans-serif;"
+    resultsDiv.style.visibility = "hidden";
+    resultsDiv.style.padding = "0px";
+    return resultsDiv
+}
```
Comment (wojm) [inline overlay.js:236]: itd be worth trying to get these into html for easier readability

File: overlay.js [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+    let index;
+    if (focusedElement.isSameNode(getSearchInputElement())) {
+        document.querySelector('.skater-result-0').focus();
+        index = 0;
+    } else {
+        // move to next search result
+        const indexOfLastFocus = focusedElement.getAttribute('class').split('-');
+        index = parseInt(indexOfLastFocus[indexOfLastFocus.length - 1]) + 1;
+        document.querySelector(`.skater-result-${index}`).focus();
+    }
```
Comment (wojm) [inline overlay.js:111]: this if block should be in the func and return the index

Comment (wojm) [reply overlay.js:111]: also the only diff between these two functions is the +1 vs -1

File: overlay.js [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+        const focusedElement = document.querySelector('.focused');
+        switch(documentEvent.key) {
+            case "Up":
+            case "ArrowUp":
+                // move to last search result or input
+                // FIXME functionize
+                if (focusedElement.getAttribute('class') === 'skater-link skater-result-0 focused') {
+                    focusInput();
+                    // FIXME: implement preventDefault better
+                    documentEvent.preventDefault();
```
Comment (wojm) [inline overlay.js]: probably want this moved up to as high as it can be. likely all the way to the top

File: overlay.js [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+                if (documentEvent.ctrlKey){
+                    // open in same window
+                } else {
+                    console.log(selectedResult);
+                    window.open(selectedResult.href);
+                    destroyOverlay();
+            }
+        }
+    }
+    if (documentEvent.key === "Escape") {
```
Comment (wojm) [inline overlay.js:43]: this has similar logic to what your switch is but is doing so seperately. probably want to combine similar logic

File: overlay.js [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+
+function setUpInputEventListener() {
+    const searchInput = getSearchInputElement();
+    searchInput.addEventListener('keyup', async function(inputEvent) {
+        const query = searchInput.value;
+        const bookmarkSearchResults = await searchBookmarks(query);
+        // refine results
+        if (Array.isArray(bookmarkSearchResults) && isValidInputEvent(inputEvent.key)) {
+            const refinedResults = refineResults(bookmarkSearchResults, query);
+            nodes = updateSearchText(refinedResults);
```
Comment (wojm) [inline overlay.js]: bas name for function. updateSearchText should not return. also doesn't look like nodes is used

File: overlay.js [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+        return true
+    });
+}
+
+function isValidInputEvent(key) {
+    const isAlphabetical = (key >= "a" && key <= "z");
+    const isNumeric = (key >= "0" && key <= "9");
+    const isBackspace = (key === "Backspace");
+    const isEnter = (key === "Enter");
+    return isAlphabetical || isNumeric || isBackspace || isEnter
```
Comment (wojm) [inline overlay.js:73]: dont need to compute all of these values. look up short circuiting

File: overlay.js [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+function moveUpOneResult() {
+    const focusedElement = document.querySelector('.focused');
+    const indexOfLastFocus = focusedElement.getAttribute('class').split(' ')[1].split('-');
+    const index = parseInt(indexOfLastFocus[indexOfLastFocus.length - 1]) - 1;
```
Comment (wojm) [inline overlay.js]: this logic is so similar except for '-1'

Comment (wojm) [reply overlay.js]: to below

File: overlay.js [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+    if (classString.length > 2) {
+        return rootClassList.slice(0, 2).join(' ');
+    } else {
+        return classString
+    }
+}
+
+function giveElementFocusedClass(index) {
+    const focusedElement = document.querySelector(`.skater-result-${index}`);
+    focusedElement.setAttribute('class', `skater-link skater-result-${index} focused`)
```
Comment (wojm) [inline overlay.js]: focused may be affected by css on page. prob try `skater-focused`

**Review by wojm — COMMENTED** [review]
stash

File: overlay.js (hunk 11 above)
Comment (wojm) [reply overlay.js:236]: maybe web components

File: overlay.js [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+    listElement.style.padding = '4px 0px 4px 0px';
+}
+
+function createListItem(result, index) {
+    const listElement = document.createElement('div');
+    const listURL = document.createElement('a');
+    const listIMG = document.createElement('img');
+
+    var matches = result.url.match(/^https?\:\/\/(?:www\.)?([^\/?#]+)(?:[\/?#]|$)/i);
+    var domain = matches && matches[1] // domain is null if no matches found
```
Comment (wojm) [inline overlay.js]: dont use var

Comment (wojm) [reply overlay.js]: also the regex is a great candidate for something to test

**Review by wojm — COMMENTED** [review]
review

File: background.js (hunk 1 above)
Comment (Ecalzo) [reply background.js]: removed

File: overlay.js (hunk 2 above)
Comment (Ecalzo) [reply overlay.js]: removed

File: overlay.js (hunk 13 above)
Comment (Ecalzo) [reply overlay.js]: @wojm I've reduced the number of times this is call to 1, only when a user hits the up arrow will it fire. Hoisting it to the top of the event listener broke the ext. during testing.

File: overlay.js (hunk 14 above)
Comment (Ecalzo) [reply overlay.js:43]: added to the switch statement

File: overlay.js (hunk 16 above)
Comment (Ecalzo) [reply overlay.js:73]: All the short circuiting results on google look like this, kinda confused

File: overlay.js (hunk 17 above)
Comment (Ecalzo) [reply overlay.js]: simplified significantly 👍

File: overlay.js (hunk 15 above)
Comment (Ecalzo) [reply overlay.js]: fix'd

Comment (Ecalzo) [reply overlay.js]: renamed function to `refreshActiveSearchResults`

**Pull request merged**